### PR TITLE
(GH-122) Support for Code Actions

### PR DESCRIPTION
**Pull request by mholo65**
Support for Code Actions, e.g. open up link which describes the diagnostics.

## Description
Added handler for Code Actions

## Related Issue
#122 

## Motivation and Context
@gep13 asked for a PoC

## How Has This Been Tested?
Works on my machine™

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/Chocolatey.Language.Server/Validations/NuspecRuleBase.cs [hunk 1]
```diff
@@ -155,18 +165,21 @@ protected Diagnostic CreateDiagnostic(SyntaxNode syntaxNode, DiagnosticSeverity
         /// <param name="severity">The Diagnostic severity to show the user.</param>
         /// <param name="message">The message to show the user.</param>
         /// <param name="wikiUrl">The wiki URL for the current rule.</param>
+        /// <param name="code">The code for the current rule.</param>
         /// <returns>The created diagnostic.</returns>
-        protected virtual Diagnostic CreateDiagnostic(int start, int end, DiagnosticSeverity severity, string message, string wikiUrl)
+        protected virtual Diagnostic CreateDiagnostic(int start, int end, DiagnosticSeverity severity, string message, string wikiUrl, string code)
```
Comment (mholo65) [inline src/Chocolatey.Language.Server/Validations/NuspecRuleBase.cs]: You probably want to have a Lookup table for Diagnostic Code => Uri to avoid passing both Code and Uri (note that Code can be `string` or `long` (https://github.com/OmniSharp/csharp-language-server-protocol/blob/0832add57b942fada096b8bfe0c266e87d84cfe3/src/Protocol/Models/DiagnosticCode.cs).

File: src/extension.ts [hunk 2]
```diff
@@ -18,12 +18,15 @@ const languageServerPaths = [
...
-    commands.registerCommand("chocolatey.apikey", () => execute("apikey"));
+    context.subscriptions.push(
+        commands.registerCommand("chocolatey.new", (arg: any) => execute("new", arg)),
+        commands.registerCommand("chocolatey.pack", () => execute("pack")),
+        commands.registerCommand("chocolatey.delete", () => deleteNupkgs()),
+        commands.registerCommand("chocolatey.push", () => execute("push")),
+        commands.registerCommand("chocolatey.installTemplates", () => execute("installTemplates")),
+        commands.registerCommand("chocolatey.apikey", () => execute("apikey")),
+        commands.registerCommand("chocolatey.open", async (uri: string) => await commands.executeCommand('vscode.open', Uri.parse(uri)))
+    );
```
Comment (mholo65) [inline src/extension.ts]: Be a good citizen and make sure these are disposed.

File: src/extension.ts [hunk 3]
```diff
@@ -158,7 +161,7 @@ function execute(cmd?: string | undefined, arg?: any[] | undefined): Thenable<st
     let ecmd: any = chocolateyManager[cmd];
     if (typeof ecmd === "function") {
         try {
-            ecmd.call(chocolateyManager, arg);
+            ecmd.call(chocolateyManager, arg);
```
Comment (mholo65) [inline src/extension.ts]: dunno why this got changed?

File: src/Chocolatey.Language.Server/CodeActionHandler.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            };
+        }
+
+        public Task<CommandOrCodeActionContainer> Handle(CodeActionParams request, CancellationToken cancellationToken)
+        {
+            var container = request.Context.Diagnostics?
+                .Where(diagnostic => diagnostic.Source.Equals("chocolatey", StringComparison.OrdinalIgnoreCase) &&
+                                     diagnostic.Code.IsString && !string.IsNullOrWhiteSpace(diagnostic.Code.String))
+                .Select(diagnostic => {
+                    var url = $"https://github.com/chocolatey/package-validator/wiki/{diagnostic.Code.String}";
```
Comment (mholo65) [inline src/Chocolatey.Language.Server/CodeActionHandler.cs]: You probably want to have a Lookup table for Diagnostic Code => Uri to avoid building the Uri here (note that Code can be `string` or `long` (https://github.com/OmniSharp/csharp-language-server-protocol/blob/0832add57b942fada096b8bfe0c266e87d84cfe3/src/Protocol/Models/DiagnosticCode.cs).

File: src/extension.ts (hunk 3 above)
Comment (AdmiringWorm) [reply src/extension.ts]: New line change. Changed from LF to CRLF.

File: src/Chocolatey.Language.Server/Validations/NuspecRuleBase.cs (hunk 1 above)
Comment (AdmiringWorm) [reply src/Chocolatey.Language.Server/Validations/NuspecRuleBase.cs]: If we implement a lookup table, we could most likely get rid of the wikirUrl parameter as we could just look up the code.

Comment (mholo65) [reply src/Chocolatey.Language.Server/Validations/NuspecRuleBase.cs]: Yes, that was my intention.

**Comment by AdmiringWorm** [comment]
Just took it for a test run, and I think it's awesome.

Was surprised that the Code Action worked in the Problems section of vscode though. (Pleasantly surprised).

code-action

**Comment by gep13** [comment]
@AdmiringWorm it really is a great PR, and I am looking forward to getting this merged into the code base, so that we can start using it across all the other rules/diagnostics.

**Comment by AdmiringWorm** [comment]
@gep13 I agree. Not that every rule have a corresponding wiki page though, but 99% do. (The SSL capability guideline don't).

**Comment by gep13** [comment]
If there isn't a wiki page for something that we are adding into this, I think we should get one created, just so that we can provide more information about the rule/diagnostic.

**Comment by mkevenaar** [comment]
@gep13 @mholo65 @AdmiringWorm I think this is in total contrast of #117

**Comment by AdmiringWorm** [comment]
@gep13 Maybe those wiki rules should be in this repo then?

Under the same ids as suggested by @mkevenaar in #117?

At least the ones without a corresponding page on the package validator?

**Comment by mkevenaar** [comment]
@AdmiringWorm IMO if we cook up a rule here, ultimately it should also become a rule in the package-validator.

**Comment by AdmiringWorm** [comment]
@mkevenaar while I don't really disagree with that, there could potentially be rules here that doesn't necessarily need to be added to the package validator (because perhaps they are handled in a different way). Just saying.

Just did a quick write-up (as a test on how it looks) for a single rule, and got the following:

image

May perhaps not need the `CHOCO00001` though, if using categories is acceptable.

**Comment by mkevenaar** [comment]
@AdmiringWorm  I like that, with perhaps one addition: Add a link to the original package-validator wiki link, if it exists.

**Comment by gep13** [comment]
>@mkevenaar said...
I think this is in total contrast of #117

Can you elaborate on this?  I don't see how they are in contrast, but rather working towards the same end goal.

**Comment by mkevenaar** [comment]
> > @mkevenaar said...
> > I think this is in total contrast of #117
> 
> Can you elaborate on this? I don't see how they are in contrast, but rather working towards the same end goal.

Let me rephrase that: I think it is a great start but based on #117 we might want to use the `CHOCO####` keys.

**Comment by gep13** [comment]
>@mkevenaar said...
Let me rephrase that: I think it is a great start but based on #117 we might want to use the `CHOCO####` keys.

Gotcha.  Yes, in which case we are on the same page.  That is one of the things that I was going to do as a result of reviewing/merging that PR.  There are a couple of things that need to be addressed in it.

**Comment by gep13** [comment]
@mholo65 I have done some small refactoring (likely still needs some work) and merged your work here:

https://github.com/gep13/chocolatey-vscode/commit/6c6943b09597813d3a9d497a24f76a5e92aeab0c

**Comment by gep13** [comment]
@mholo65 your changes have been merged, thanks for your contribution 👍

**Comment by gep13** [comment]
@all-contributors please add @mholo65 for code, ideas, review
